math/n_choose_r.cpp

```diff
@@ -0,0 +1,82 @@
+/**
+ * @file
+ * @brief [Combinations](https://en.wikipedia.org/wiki/Combination) n choose r function implementation
+ * @details
+ * A very basic and efficient method of calculating
+ * choosing r from n different choices.
+ * \f$ \binom{n}{r} = \frac{n!}{r! (n-r)!} \f$
+ *
+ * @author [Tajmeet Singh](https://github.com/tjgurwara99)
+ */
+
+#include <iostream>   /// for io operations
+#include <cassert>   /// for assert
+
+
+/**
+ * @namespace math
+ * @brief Mathematical algorithms
+ */
+namespace math {
+/**
+ * @brief This is the function implementation of \f$ \binom{n}{r} \f$
+ * @details
+ * We are calculating the ans with iterations
+ * instead of calculating three different factorials.
+ * Also, we are using the fact that 
+ * \f$ \frac{n!}{r! (n-r)!} = \frac{(n - r + 1) \times \cdots \times n}{1 \times \cdots \times r} \f$
+ * @tparam T Only for integer types such as long, int_64 etc
+ * @param n \f$ n \f$ in \f$ \binom{n}{r} \f$
+ * @param r \f$ r \f$ in \f$ \binom{n}{r} \f$
+ * @returns ans \f$ \binom{n}{r} \f$
+ */
+template <class T>
+T n_choose_r(T n, T r) {
+    if(r > n / 2)
+        r = n - r; // Because of the fact that  nCr(n, r) == nCr(n, n - r)
+    T ans = 1;
+    for(int i = 1; i <= r; i++) {
+        ans *= n - r + i;
+        ans /= i;
+    }
+    return ans;
+}
```

Comment:
I think it makes more sense to use template specilization in this case, because that will automatically ensure that parameters are nothing other than `int` or `int64_t` or `long` whereas in this implementation there will not be any error if someone passes parameters of type `double` or something else rather it will produce wrong results.

Reply:
I think specialisation wouldn't make sense in this case because we don't want a different behaviour for other datatype but rather we want them to nothing if its anything other than integer types - overloading would make more sense. If I were to specialise this function it would very much look like the following (this is the only way I can think of but if you have suggestions, please let me know):

```C++
template <class T>
T n_choose_r(T n, T r) {
    // this should do nothing because we want to specialise to int types
}

template <>
int_64 n_choose_r(int_64 n, int_64 r) {
    if(r > n / 2)
        r = n - r; // Because of the fact that  nCr(n, r) == nCr(n, n - r)
    int_64 ans = 1;
    for(int i = 1; i <= r; i++) {
        ans *= n - r + i;
        ans /= i;
    }
    return ans;
}


```
and so on for other types which makes the idea of having templates redundant (unless you have a better way - if so please let me know as it would help me understand specialisation better). Also consider the case that we have a custom BigInt library and addition, subtraction, multiplication & division is defined for them then this would not work.

Let me know what you think about it😄 I will try to implement what you think would be best.

Reply:
yes you are right using template specilization leads to redundant code. A better way to go with this is to use some sort of functionality that will `ON` or `OFF`  the single general template based on some sort of condition (condition is based on the types of parameters passed in this case) and to achieve this we can have two ways, one is `tag dispatch` and other is `SFINAE`. You can use any one of them.

Reply:
But i think this would make this simple program quite complex for beginners to understand. isn't it?

Reply:
I quite agree, it's meant to be a very simple yet efficient way of calculating nCk and adding both `tag dispatch` and `substitution failure not an error` would indeed just make it very complicated to understand for beginners. If TheAlgorithms collaborative is working on a library then it may be useful for converting this into a STL like function using `tag dispatcher` but if not then I don't think we should do anything to it. What do you think @ayaankhan98?

Reply:
yes the soul purpose of these implementaions is to provide a educational medium so there is no need to make simple things tough, better to leave it as it is.

Reply:
@tjgurwara99 Thanks for your contribution

Reply:
Also, do you have any idea why the `gh-pages` for my fork of this repo is having trouble with graphviz? for example, if you go to the [tjgurwara99.github.io/C-Plus-Plus/n_choose_r](https://tjgurwara99.github.io/C-Plus-Plus/d1/dbb/n__choose__r_8cpp.html#details) you would see that there is an error on it. Any idea of what that is about? 

I ask this because when I compile this same code in the gitpod above it just gives me correct html file with no errors so was just wondering if I have to change something in the Doxygen workflow or something.